library/psa_crypto_se.c

```diff
@@ -0,0 +1,79 @@
+/*
+ *  PSA crypto support for secure element drivers
+ */
+/*  Copyright (C) 2019, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of Mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_PSA_CRYPTO_C)
+
+#include <string.h>
+
+#include "psa_crypto_se.h"
+
+typedef struct
+{
+    psa_key_lifetime_t lifetime;
+    const psa_drv_se_t *methods;
+} method_table_entry_t;
+
+static method_table_entry_t driver_table[PSA_MAX_SE_DRIVERS];
+
+psa_status_t psa_register_se_driver(
+    psa_key_lifetime_t lifetime,
+    const psa_drv_se_t *methods)
+{
+    size_t i;
+
+    if( methods->hal_version != PSA_DRV_SE_HAL_VERSION )
+        return( PSA_ERROR_NOT_SUPPORTED );
+    if( lifetime == PSA_KEY_LIFETIME_VOLATILE ||
+        lifetime == PSA_KEY_LIFETIME_PERSISTENT )
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    for( i = 0; i < PSA_MAX_SE_DRIVERS; i++ )
+    {
+        if( driver_table[i].lifetime == 0 )
```

Comment:
`lifetime` equal to 0 (coincidentally?) is also equal to PSA_KEY_LIFETIME_VOLATILE - is it intended that an unused or uninitialized value uses it? If so, I'd suggest changing 0 to PSA_KEY_LIFETIME_VOLATILE, if not - introducing a separate define for it.

Reply:
I think that @AndrzejKurek raises an important point: What does `lifetime == 0` mean?

I think that according to this code 0 is unititialized, but that value also means PSA_KEY_LIFETIME_VOLATILE. Personally, I really dislike this kind of coincidence. Why not have a dedicated value that means "unitialized/free/unused" driver table entry?

Reply:
“Unused” is 0 from static initialization. 0 is invalid due to being the value of `PSA_KEY_LIFETIME_VOLATILE`. This is not a mere coincidence: 0 is an “invalid” or “safe default” value wherever practical in the API. I don't see what would be gained by adding a separate field to indicate whether an entry is initialized.

Reply:
I think Andres was suggesting a different value, not a different field, and to replace `lifetime == 0` with `lifetime == PSA_KEY_LIFETIME_INVALID`.

Reply:
Static initialization makes the initial value 0. It isn't possible to choose a different initial value that's different from `PSA_KEY_LIFETIME_VOLATILE`.

0 is `VOLATILE`, not `INVALID`, because it's a safe default for a key. If you don't explicitly say where your key is, it stays in RAM. I don't see a reason to make 0 `PSA_KEY_LIFETIME_INVALID`, which would require applications to set the key's lifetime explicitly every time they create a key.